File: part2-simple-linear-regression/slr.py
# Data preprocessing templace:
# 1. Import the libraries
# 2. Import the dataset
# 3. Check for missing values
# 4. Encode discrete (categorical) values
# 5. CHeck if feature scaling is necessary
# 6. Split the dataset in training set and test set

# Section one: import the needed libraries
import pandas as pd # This package is great for data handling in python

# Section two: import the dataset
dataset = pd.read_csv("/home/felipe/Documentos/Machine Learning A-Z/"
	"Part 2 - Regression/Section 4 - Simple Linear Regression/Salary_Data.csv")
p_input = dataset.iloc[:, :-1].values # Independent variable 
p_output = dataset.iloc[:, -1].values # Dependent variable

# Section three: check for missing data
# No missing data!

# Section four: Encode categorical values
# No categorical values!

# Section five: check if feature scaling is necessary
# It can be, depending of the package used.

# Section six: split the dataset into train and test set
from sklearn.model_selection import train_test_split as tts
p_input_train, p_input_test, p_output_train, p_output_test = tts(p_input, p_output, test_size = 0.33)

# Fitting Simple Linear Regression (SLR) into the training set
from sklearn.linear_model import LinearRegression as lr
regressor = lr() # Call the constructor method of the sklearn linear regressor
regressor.fit(p_input_train, p_output_train) # Construct the model with the train set

# Testing the model with the test set
predictions = regressor.predict(p_input_test) # Make predictions

# Plot in the train set
import matplotlib.pyplot as plt
plt.scatter(p_input_train, p_output_train, color = 'red', label = 'Train Set')
plt.scatter(p_input_test, p_output_test, color = 'green', label = 'Test Set')
# plt.ticklabel_format(style = 'sci', axis = 'y', scilimits = (0, 0))
plt.plot(p_input_train, regressor.predict(p_input_train), color = 'blue')
plt.title('Salary vs Experience')
plt.xlabel('Experience')
plt.ylabel('Salary')
plt.legend()
plt.show()

# Be careful! When plotting the TEST SET, the linear regression line
# is the exactly same line of the previous plot!

File: testing-section/Regression-testing/part2-simple-linear-regression/slr.R
# Data preprocessing templace:
# 1. Import the dataset
# 2. Check for missing values
# 3. Encode discrete (categorical) values
# 4. CHeck if feature scaling is necessary
# 5. Split the dataset in training set and test set

# Section one: import the dataset
dataset <- read.csv("/home/felipe/Documentos/Machine Learning A-Z/Part 2 - Regression/Section 4 - Simple Linear Regression/Salary_Data.csv")

# A simple line
# y = a + bx
# a: linear coef
# b: angular coef (slope)

# Section 2: No missing values!

# Section 3: No categorical values!

# Section 4: No scaling necessary, on this case.

# Section 5: Split the dataset into training and test set
library(caTools)
split <- sample.split(dataset$YearsExperience, SplitRatio = 2/3)
train_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

# Construct the Linear Regression Model
regressor <- lm(formula = Salary ~ YearsExperience, data = train_set)
summary(regressor)

# Using the linear model into the test set
predictions <- predict(object = regressor, newdata = test_set)

# Plot the results
library(ggplot2)
ggplot() + 
	geom_point(aes(x = train_set$YearsExperience, y = train_set$Salary), 
		colour = 'blue') +
	geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary), 
		colour = 'green') +
	geom_line(aes(x = train_set$YearsExperience, y = predict(object = regressor, newdata = train_set)),
		colour = 'red') +
	ggtitle('Salary vs Experience') +
	xlab('Years of Experience') +
	ylab('Salary')